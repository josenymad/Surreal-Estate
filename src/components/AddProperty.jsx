import React, { useState } from "react";
import "../styles/add-property.css";
import postPropertyDetails from "../requests/postProperty";
import Alert from "./Alert";

const AddProperty = () => {
  const initialState = {
    fields: {
      title: "",
      type: "",
      bedrooms: "",
      bathrooms: "",
      price: "",
      city: "",
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddProperty = (event) => {
    event.preventDefault();
    setAlert({ message: "", isSuccess: false });
    postPropertyDetails(fields, setAlert);
  };
  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  return (
    <div className="add-property">
      <h3 className="add-property__header">Tell us about your new property:</h3>
      <form onSubmit={handleAddProperty} className="add-property__form">
        <label htmlFor="title">
          Property Title:
          <input
            id="title"
            name="title"
            value={fields.title}
            onChange={handleFieldChange}
            placeholder="2 bedroom flat"
            className="form__input"
          />
        </label>
        <label htmlFor="type">
          Property Type:
          <select
            id="type"
            name="type"
            value={fields.type}
            onChange={handleFieldChange}
            className="form__input"
          >
            <option value="">Please select</option>
            <option value="Flat">Flat</option>
            <option value="Detached">Detached</option>
            <option value="Semi-Detached">Semi-Detached</option>
            <option value="Terraced">Terraced</option>
            <option value="End of Terrace">End of Terrace</option>
            <option value="Cottage">Cottage</option>
            <option value="Bungalow">Bungalow</option>
          </select>
        </label>
        <label htmlFor="bedrooms">
          Number of bedrooms:
          <input
            id="bedrooms"
            name="bedrooms"
            value={fields.bedrooms}
            onChange={handleFieldChange}
            placeholder="2"
            className="form__input"
          />
        </label>
        <label htmlFor="bathrooms">
          Number of bathrooms:
          <input
            id="bathrooms"
            name="bathrooms"
            value={fields.bathrooms}
            onChange={handleFieldChange}
            placeholder="1"
            className="form__input"
          />
        </label>
        <label htmlFor="price">
          Property value:
          <input
            id="price"
            name="price"
            value={fields.price}
            onChange={handleFieldChange}
            placeholder="150000"
            className="form__input"
          />
        </label>
        <label htmlFor="city">
          City:
          <select
            id="city"
            name="city"
            value={fields.city}
            onChange={handleFieldChange}
            className="form__input"
          >
            <option value="">Please select</option>
            <option value="Manchester">Manchester</option>
            <option value="Leeds">Leeds</option>
            <option value="Sheffield">Sheffield</option>
            <option value="Liverpool">Liverpool</option>
          </select>
        </label>
        <label htmlFor="email">
          Email:
          <input
            id="email"
            name="email"
            value={fields.email}
            onChange={handleFieldChange}
            placeholder="contact@email.com"
            className="form__input"
          />
        </label>
        <button type="submit" className="add-property__form__button">
          Add
        </button>
        <Alert
          message={alert.message}
          success={alert.isSuccess}
          className="alert"
        />
      </form>
    </div>
  );
};

export default AddProperty;
