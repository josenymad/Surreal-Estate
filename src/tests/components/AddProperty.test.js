import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AddProperty from "../../components/AddProperty";

describe("Add Property", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders 'Tell us about your new property:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const header = screen.getByText("Tell us about your new property:");

    expect(header).toBeInTheDocument();
  });

  it("renders 'Property Title:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const propertyTitleLabel = screen.getByLabelText("Property Title:");

    expect(propertyTitleLabel).toBeInTheDocument();
  });

  it("renders 'Property Type:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const propertyTypeLabel = screen.getByLabelText("Property Type:");

    expect(propertyTypeLabel).toBeInTheDocument();
  });

  it("renders 'Number of bedrooms:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const numberOfBedroomsLabel = screen.getByLabelText("Number of bedrooms:");

    expect(numberOfBedroomsLabel).toBeInTheDocument();
  });

  it("renders 'Number of bathrooms:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const numberOfBathroomsLabel = screen.getByLabelText(
      "Number of bathrooms:"
    );

    expect(numberOfBathroomsLabel).toBeInTheDocument();
  });

  it("renders 'City:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const cityLabel = screen.getByLabelText("City:");

    expect(cityLabel).toBeInTheDocument();
  });

  it("renders 'Email:'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const emailLabel = screen.getByLabelText("Email:");

    expect(emailLabel).toBeInTheDocument();
  });

  it("has a button", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const buttons = screen.getAllByRole("button");

    expect(buttons).toHaveLength(1);
  });

  it("button has the text 'Add'", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const button = screen.getByRole("button");

    expect(button).toHaveTextContent("Add");
  });
});
