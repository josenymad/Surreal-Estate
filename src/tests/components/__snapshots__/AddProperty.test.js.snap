// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Property renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <h3
      class="add-property__header"
    >
      Tell us about your new property:
    </h3>
    <form
      class="add-property__form"
    >
      <label
        for="title"
      >
        Property Title:
        <input
          class="form__input"
          id="title"
          name="title"
          placeholder="2 bedroom flat"
          value=""
        />
      </label>
      <label
        for="type"
      >
        Property Type:
        <select
          class="form__input"
          id="type"
          name="type"
        >
          <option
            value=""
          >
            Please select
          </option>
          <option
            value="Flat"
          >
            Flat
          </option>
          <option
            value="Detached"
          >
            Detached
          </option>
          <option
            value="Semi-Detached"
          >
            Semi-Detached
          </option>
          <option
            value="Terraced"
          >
            Terraced
          </option>
          <option
            value="End of Terrace"
          >
            End of Terrace
          </option>
          <option
            value="Cottage"
          >
            Cottage
          </option>
          <option
            value="Bungalow"
          >
            Bungalow
          </option>
        </select>
      </label>
      <label
        for="bedrooms"
      >
        Number of bedrooms:
        <input
          class="form__input"
          id="bedrooms"
          name="bedrooms"
          placeholder="2"
          value=""
        />
      </label>
      <label
        for="bathrooms"
      >
        Number of bathrooms:
        <input
          class="form__input"
          id="bathrooms"
          name="bathrooms"
          placeholder="1"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Property value:
        <input
          class="form__input"
          id="price"
          name="price"
          placeholder="150000"
          value=""
        />
      </label>
      <label
        for="city"
      >
        City:
        <select
          class="form__input"
          id="city"
          name="city"
        >
          <option
            value=""
          >
            Please select
          </option>
          <option
            value="Manchester"
          >
            Manchester
          </option>
          <option
            value="Leeds"
          >
            Leeds
          </option>
          <option
            value="Sheffield"
          >
            Sheffield
          </option>
          <option
            value="Liverpool"
          >
            Liverpool
          </option>
        </select>
      </label>
      <label
        for="email"
      >
        Email:
        <input
          class="form__input"
          id="email"
          name="email"
          placeholder="contact@email.com"
          value=""
        />
      </label>
      <button
        class="add-property__form__button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
